### PR TITLE
feat: add type predicate support for `roomPos.findxxx`

### PR DESCRIPTION
**Pull request by DiamondMofeng**
### Mention

I don't know if this is a breaking change as I have changed the internal type `filterOption`. Anyway, as we are going to release next major version to enable strict id check, we can also merge this as a part of 4.0.0.

### Brief Description

here is a util function to generate a filter callback with type predicate
```typescript
const isStructureType = <T extends StructureConstant>(structureType: T) => {
    return (structure: AnyStructure): structure is ConcreteStructure<T> => {
        return structure.structureType === structureType;
    };
};
```

before this pr:

```typescript
    {
        let towers = creep.pos.findInRange([] as AnyStructure[], 2, {
            filter: isStructureType(STRUCTURE_TOWER),
        });
        towers[0].attack(creep);    //! ERROR! type of towers is `AnyStructure[]` instead of `StructureTower[]`
    }
    {
        let tower = creep.pos.findClosestByPath([] as AnyStructure[], {
            filter: isStructureType(STRUCTURE_TOWER),
        });
        tower?.attack(creep);    //! ERROR! type of towers is `AnyStructure` instead of `StructureTower`
    }
    {
        let tower = creep.pos.findClosestByRange([] as AnyStructure[], {
            filter: isStructureType(STRUCTURE_TOWER),
        });
        tower?.attack(creep);    //! ERROR! type of towers is `AnyStructure` instead of `StructureTower`
    }
```

after this pr:

no error occurres. These `tower` variables are correctly typed as `StructureTower`

### Checklists

- [x] Test passed
- [x] Coding style (indentation, etc)
- [x] Edits have been made to `src/` files not `index.d.ts`
- [x] Run `npm run dtslint` to update `index.d.ts`

### CODE REVIEW

File: dist/index.d.ts [hunk 1]
```diff
@@ -2174,12 +2174,15 @@ type FIND_RUINS = 123;
 
 // Filter Options
 
-interface FilterOptions<T extends FindConstant, S extends FindTypes[T] = FindTypes[T]> {
-    filter: FilterFunction<FindTypes[T], S> | FilterObject | string;
+interface PredicateFilterOptions<T, S extends T> {
```
Comment (DiamondMofeng) [inline dist/index.d.ts:2177]: would anyone give this a better name?

File: dist/index.d.ts [hunk 2]
```diff
@@ -3880,68 +3883,92 @@ interface RoomPosition {
      * @param opts An object containing pathfinding options (see Room.findPath), or one of the following: filter, algorithm
      * @returns An instance of a RoomObject.
      */
-    findClosestByPath<K extends FindConstant, S extends FindTypes[K]>(
+    findClosestByPath<K extends FindConstant, T extends FindTypes[K], S extends T>(
         type: K,
-        opts?: FindPathOpts & Partial<FilterOptions<K, S>> & { algorithm?: FindClosestByPathAlgorithm },
+        opts?: FindPathOpts & Partial<PredicateFilterOptions<T, S>> & { algorithm?: FindClosestByPathAlgorithm },
     ): S | null;
```
Comment (DiamondMofeng) [inline dist/index.d.ts:3889]: detailed type should be placed on top

File: dist/index.d.ts [hunk 3]
```diff
@@ -3880,68 +3883,92 @@ interface RoomPosition {
...
-    findClosestByPath<S extends AnyStructure>(
+    findClosestByPath<K extends FindConstant, T extends FindTypes[K] = FindTypes[K]>(
+        type: K,
+        opts?: FindPathOpts & Partial<FilterOptions<FindTypes[K]>> & { algorithm?: FindClosestByPathAlgorithm },
+    ): T | null;
+    findClosestByPath<T extends AnyStructure>(
```
Comment (DiamondMofeng) [inline dist/index.d.ts:3894]: I would rather delete this declaration than keep this. But I have seen many player who use find<StructureFoobar>, so keep this for compatibility

File: dist/index.d.ts [hunk 4]
```diff
@@ -3880,68 +3883,92 @@ interface RoomPosition {
...
         type: FIND_STRUCTURES | FIND_MY_STRUCTURES | FIND_HOSTILE_STRUCTURES,
-        opts?: FindPathOpts & Partial<FilterOptions<FIND_STRUCTURES, S>> & { algorithm?: FindClosestByPathAlgorithm },
-    ): S | null;
+        opts?: FindPathOpts & Partial<FilterOptions<AnyStructure>> & { algorithm?: FindClosestByPathAlgorithm },
```
Comment (DiamondMofeng) [inline dist/index.d.ts:3896]: Maybe we can infer this `AnyStructure` from the first argument, but another type parameter would be introduced.

File: dist/index.d.ts [hunk 5]
```diff
@@ -2174,12 +2174,15 @@ type FIND_RUINS = 123;
...
+interface PredicateFilterOptions<T, S extends T> {
+    filter: PredicateFilterFunction<T, S>;
+}
+interface FilterOptions<T> {
+    filter: FilterFunction<T> | FilterObject | string;
 }
 
-// We do not need to mark params as optional, because this is used for callback functions, whose params are always optional
-type FilterFunction<T, S extends T> = (object: T, index: number, collection: T[]) => object is S;
+type PredicateFilterFunction<T, S extends T> = (object: T, index: number, collection: T[]) => object is S;
```
Comment (Jomik) [inline dist/index.d.ts:2184]: I think this is redundant. You can just pass `T` in for the case where `S` is not given.

File: dist/index.d.ts (hunk 1 above)
Comment (Jomik) [reply dist/index.d.ts:2177]: We could probably avoid it, which I would argue we should.
There is already a default for FilterOptions, so you just need to add an extra, optional, parameter to the `find*` methods

**Review by xilexio — APPROVED** [review]
Looks good to me. See attached comments.

File: dist/index.d.ts (hunk 5 above)
Comment (xilexio) [reply dist/index.d.ts:2184]: > I think this is redundant. You can just pass `T` in for the case where `S` is not given.

I think leaving `FilterFunction` is more about making sure that just any lodash filter (that returns `unknown` may be used).

Comment (Jomik) [reply dist/index.d.ts:2184]: That should be unrelated :)

File: dist/index.d.ts (hunk 4 above)
Comment (DiamondMofeng) [reply dist/index.d.ts:3896]: > Is this about the declaration starting on line 3984? You want to infer the type of application of `findClosestByPath` from the type of first argument of function defined in `filter`? Wouldn't that be difficult due to `FilterOption.filter` being of a union type? Isn't it already covered by the version from line 3886?

I mean inferring the `AnyStructure` from the given `FIND_(MY)_STRUCTURE`. 
TS throws an error in the case below.
```
{
    let towers = creep.pos.findClosestByPath<StructureTower>(FIND_MY_STRUCTURES, {
        filter: (s) => s?.my // ERROR: `my` is not exist on `AnyStructure`
    });
}
```

Comment (Jomik) [reply dist/index.d.ts:3896]: You could do `F extends FIND_STRUCTURES | FIND_MY_STRUCTURES | FIND_HOSTILE_STRUCTURES` and then use `FindTypes[F]` to get the structure type, I think. Or at least a similar idea to that with a map.

File: dist/index.d.ts (hunk 5 above)
Comment (DiamondMofeng) [reply dist/index.d.ts:2181]: I have merged `PredicateFilterOption` into `FilterOption` now

File: dist/index.d.ts (hunk 4 above)
Comment (DiamondMofeng) [reply dist/index.d.ts:3889]: Let's keep it as is now.

File: dist/index.d.ts (hunk 3 above)
Comment (DiamondMofeng) [reply dist/index.d.ts:3887]: Let's keep it as is now.

File: dist/index.d.ts (hunk 1 above)
Comment (DiamondMofeng) [reply dist/index.d.ts]: `PredicateFilterOptions` has been merged into `FilterOption` now. But there is still a `PredicateFilterFunction`. Let's keep it as-is now. Feel free to give any better name.

**Pull request closed without merging**